Update DateRange.tsx

Aryam2121 opened the pull request:
Rounding Function: The roundToNearest15Minutes function rounds the selected time to the nearest 15 minutes.
Handle Change Logic: The handleChange function has been modified to use the rounding function before dispatching the updated date values.
Complete Component: The full DateRange component is included for context, with both date and datetime pickers supported.so;ve issue #1452

Aryam2121 committed:
Update DateRange.tsx

Rounding Function: The roundToNearest15Minutes function rounds the selected time to the nearest 15 minutes.
Handle Change Logic: The handleChange function has been modified to use the rounding function before dispatching the updated date values.
Complete Component: The full DateRange component is included for context, with both date and datetime pickers supported.so;ve issue #1452

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/DateRange.tsx:
```diff
@@ -6,8 +6,8 @@
  *        http://www.apache.org/licenses/LICENSE-2.0
  *
- * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
- * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
```
please do not modify head comment as it is generated and eventually replaced automatically.

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/DateRange.tsx:
```diff
@@ -54,7 +62,7 @@ const getRangeDateTime = (
             dates = JSON.parse(json);
         } catch {
-            // too bad
+            // Handle error if JSON parsing fails
```
we don't want to handle this error

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/DateRange.tsx:
```diff
@@ -86,7 +94,7 @@ const DateRange = (props: DateRangeProps) => {
             setValue((dates) => {
                 if (v !== null && isValid(v)) {
-                    const newDate = getTimeZonedDate(v, tz, withTime);
+                    const newDate = roundToNearest15Minutes(getTimeZonedDate(v, tz, withTime)); // Round to nearest 15 mins
```
this behavior should be optional and commanded by a property.
I propose that this property would be set the value in time format (hh:mm:ss) so that we can define before showing the time selector, which times the user can select
I propose the property name: time_step
In this example, `time_step=00:15` 
or `time_step=01:00` for letting the user select only time on the hour.
This should also apply to date_selector and time_selector

FredLL-Avaiga commented on frontend/taipy-gui/src/components/Taipy/DateRange.tsx:
```diff
@@ -218,20 +226,23 @@ const DateRange = (props: DateRangeProps) => {
                                 id={id && id + "-field"}
                                 className={getSuffixedClassNames(className, "-text")}
-                                width={props.width && "100%"}
+                                width={props.width}
+                                disabled
+                                label={props.labelStart}
```
I don't understand those changes.

- Why change the width ?
- Field has no `disabled` or `label` property

FredLL-Avaiga reviewed (changes requested):
Good work @Aryam2121 
I think this issue needs a bit more thinking before implementation.
In the code you propose, every component is impacted by the stepping to 15 minutes.
I think this feature should be optional and configurable and should be applied before user selection to limit the available choices.
It should also apply to date_selector and time_selector.
If you provide code, you need also to provide tests.

Aryam2121 committed:
Update DateRange.tsx

New timeStep Prop: An optional timeStep prop allows the user to define how time is rounded (e.g., to the nearest 15 minutes).
roundToStep Function: This function applies the rounding logic, adjusting the minutes of a selected time to the nearest multiple of the provided timeStep.
handleChange Function: Before dispatching the updated time, it applies the rounding logic if the timeStep prop is provided solve issue #1452 and updated code as per review #2034

FredLL-Avaiga commented:
Are you still working on this PR @Aryam2121  ?
If you are, please resolve the conflicts, update your branch and answer the questions
If you are not, please close the PR

Aryam2121 commented:
i update my Branch  already as your requirements you have to check the updated branch i want you review my updated branch

FredLL-Avaiga commented:
> i update my Branch already as your requirements you have to check the updated branch i want you review my updated branch

I'll review it once you have fixed the conflicts

Aryam2121 commented:
so you merge this please i already fixed all conflicts

FredLL-Avaiga commented:
> so you merge this please i already fixed all conflicts

![image](https://github.com/user-attachments/assets/c4803a5d-005c-450f-aacd-abcef01e33bc)

FabienLelaquais commented:
Does not do much - and certainly not enough.
